README.md

```diff
@@ -59,6 +59,58 @@ Example: Instead of `file-server.ts` use `file_server.ts`.
 More specifically, code should be wrapped at 80 columns and use 2-space
 indentation and use camel-case. Use `//format.ts` to invoke prettier.
 
+### Exported functions: max 2 args, put rest into an options object.
+
+When designing APIs, stick to the following rules.
+
+1. A function that is part of the public API takes 0-2 'normal' arguments,
+   plus (if necessary) an options object (so max 3 total).
+
+2. Optional parameters _always_ go into the options object.
```

Comment:
What about the case with one optional parameter?

Reply:
> I think this is an example where an options object makes the API more cluttered:

I disagree -- in fact this is a great example in the "BAD" category.
* It's not unlikely that we'll want to add more options at some point. Think of `append`, `exclusive`/`overwrite`, `encoding`, `owner`, `retryFail`. You name it ;)
 * Most people won't care to pass `perm`.

So in the common case it's good: `writeFileSync("foo.txt", "lorem ipsum")`;
And it allows for extensibility: `writeFileSync("foo.txt", "lorem ipsum", { append: true })`

Reply:
writeFile isn't a syscall, but it's a very basic op that I would not want to extend further. 

* `append` is in appropriate, this is for writing whole files. Append is done with `deno.open("txt", "a")`
* `exclusive/overwrite` maybe
* `encoding` i would not add this because it changes the signature (I realize one can model this dependent signatures in TS, but I find that unnecessarily complex.)
* `owner` meh - I guess - but I would really rather send people to `open` instead.
* `retryFail` this is definitely something that should be done at a higher layer.

The problem with forcing people to use objection objects when there's only one option is
1. People will be encouraged to go looking for more options, to fill this mandatory interface, which leads to API complexity; and
2. It's clearly more typing than a single optional parameter.

Reply:
> writeFileSync isn't a syscall, but it's a very basic op that I would not want to extend further.

I would argue that in that case, the `mode` parameter is probably just superfluous.

Reply:
To create a file with non default mode, use `deno.open("txt", { mode: xxx })`

Reply:
> 2\. people will be encouraged to go looking for more options, to fill this mandatory interface, which leads to API complexity.

I feel that concern too. However it's a slippery slope argument. If we don't let people add options then it won't grow.

Reply:
I would rather we not be so super prescriptive for starters, and allow people to have a single optional argument. (Also matches what already exists in deno's API.) Down the road we can tighten this if it makes sense.

Reply:
> The problem with forcing people to use option objects when there's only one option is

The problem with allowing people to add "random" optional params is that it freezes the API to support only that particular one random flag.
I for one never use the `mode` argument (probably because windows). However I think `exclusive` would be quite nice -- would much rather do `writeFileSync("foo", "lorem ipsum", { exclusive: true })` than `let fd; try { fd = open("foo", "wx+"); write(fd, "lorem ipsum"); } finally { close(fd) }

Reply:
I mean - it's a good point - but it results in more code in some cases - so I'm hesitant. 

Is there someway we can say "if you're sure there will only ever be one optional argument, it maybe be placed in the parameters directly"

Reply:
I still have a problem with "always" - I don't think it's fair to force optional arguments in some situations where it's very clear there will only ever be one optional argument. 

I understand that you're trying to allow function signatures to mature safely over time, but there are situations where a single optional argument is reasonable. And yes, sometimes even after careful consideration, there may be a case in the future when a signature with a single optional argument will need a second. In such situations we either bite the bullet and making a breaking change, or choose a new name for that function. We will not, however, get into this situation where we do complex type detection on an optional argument to determine if its using the legacy or contemporary API. So, I understand your concern, but let's not force more code where less is fine.

Reply:
ok -- try and help me out with a _good_ example where I can't think of the next parameter that one might add within 5 seconds.

Reply:
Suppose we have an http server: `server.listen(addr = ":8080")`